src/main/java/htsjdk/beta/plugin/HtsCodec.java

```diff
@@ -0,0 +1,371 @@
+ *     <ul>
+ *     <li> {@link htsjdk.beta.plugin.registry.HtsCodecResolver#resolveForDecoding(Bundle)} </li>
+ *     <li> {@link htsjdk.beta.plugin.registry.HtsCodecResolver#resolveForEncoding(Bundle)} </li>
+ *     <li> {@link htsjdk.beta.plugin.registry.HtsCodecResolver#resolveForEncoding(Bundle, HtsVersion)} </li>
+ * </ul>
+ * <H3>Formats That Use a Custom URI or Protocol Scheme</H3>
+ * <p>
+ *     Many file formats consist of a single file that resides on a file system that is supported by a
+ *     {@link java.nio} file system provider. Codecs that support such formats are generally agnostic
+ *     about the IOPath or URI protocol scheme used to identify their resources, and assume that file contents
+ *     can be accessed directly via a single stream created via a {@link java.nio} file system provider.
+ * <p>
+ *     However, some file formats use a specific, well known URI format or protocol scheme, often
+ *     to identify a remote or otherwise specially-formatted resource, such as a local database
+ *     that is distributed across multiple physical files. These codecs may bypass direct file {@link java.nio}
+ *     system access, and instead use specialized code to access their underlying resources.
+ * <p>
+ *     For example, the {@link htsjdk.beta.codecs.reads.bam.bamV1_0.BAMCodecV1_0} assumes that IOPath
+ *     resources can be accessed as a stream on a single file via either the "file://" protocol, or
+ *     other protocols such gs:// or hdfs:// that have {@link java.nio} file system providers. It does
+ *     not require or assume a particular URI format, and is agnostic about URI scheme.
+ * <p>
+ *     In contrast, the {@link htsjdk.beta.codecs.reads.htsget.htsgetBAMV1_2.HtsgetBAMCodecV1_2} codec
+ *     is a specialized codec that handles remote resources via the "http://" protocol.
+ *     It uses {@code http} to access the underlying resource, and bypasses direct {@link java.nio}
+ *     file system access.
+ * <p>
+ *     Codecs for formats that use a custom URI format or protocol scheme such as {@code htsget} must be
+ *     able to determine if they can decode or encode a resource purely by inspecting the IOPath/URI, and
+ *     should follow these guidelines:
+ *     <ul>
+ *         <li>return true when {@link #ownsURI(IOPath)} is presented with an IOPath with
+ *         a conforming URI </li>
+ *         <li>return true when {@link #canDecodeURI(IOPath)} is presented with an IOPath
+ *         with a conforming URI </li>
+ *         <li> ensure that for a given IOPath, {@link #ownsURI(IOPath)} == {@link #canDecodeURI(IOPath)}
+ *         <li>always return 0 from the {@link #getSignatureProbeLength()} method</li>
+ *         <li>always return 0 from the {@link #getSignatureLength()} method</li>
+ *     </ul>
+ * <p>
+ * <H3>Codec Implementation Guidelines</H3>
+ * <p>
+ *     <ul>
+ *         <li>
+ *             An HtsCodec class should implement only a single version of a single file format.
```

Comment:
I was thinking of the version here as the htsget version, not the version of the underlying file. It may be that we wind up having only one htsget codec for all reads that handles. It would have to in turn delegate to the correct underlying decoder.